File: tree/path_and_sum/PathSum2.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2023-01-21 14:47:56
 * @LastEditTime: 2023-01-21 15:02:31
 */
package tree.path_and_sum;

import java.util.ArrayList;
import java.util.List;

import tree.TreeNode;

/**
 * leetcode 113 middle 路径总和2
 * 
 * 给你二叉树的根节点 root 和一个整数目标和 targetSum ，找出所有 从根节点到叶子节点 路径总和等于给定目标和的路径。
 * 叶子节点 是指没有子节点的节点。
 */
public class PathSum2 {
    List<List<Integer>> ans = new ArrayList<>();

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return new ArrayList<>();
        }
        process(root, targetSum, new ArrayList<>());
        return ans;
    }

    public void process(TreeNode root, int targetSum, List<Integer> path) {
        if (root == null) {
            return;
        }

        path.add(root.val);
        targetSum -= root.val;

        if (root.left == null && root.right == null && targetSum == 0) {
            ans.add(new ArrayList<>(path));
            return;
        }

        process(root.left, targetSum, path);
        process(root.right, targetSum, path);

        path.remove(path.size()-1);
        targetSum += root.val;
    }
}

File: tree/path_and_sum/path_sum2.go
package pathandsum

import "trouble_solver/tree"

// 找出所有 从根节点到叶子节点 路径总和等于给定目标和的路径。
func PathSum(root *tree.TreeNode, targetSum int) [][]int {
	if root == nil {
		return [][]int{}
	}
	ans := make([][]int, 0)
	var process func(root *tree.TreeNode, targetSum int, curPath []int)
	process = func(root *tree.TreeNode, targetSum int, curPath []int) {
		if root == nil {
			return
		}
		curPath = append(curPath, root.Val)
		targetSum -= root.Val
		if root.Left == nil && root.Right == nil && targetSum == 0 {
			// 不能直接将currPath放到result里面, 因为currPath是共享的, 每次遍历子树时都会被修改（没太明白）
			// ans = append(ans, copySlice(curPath))
			ans = append(ans, copySlice(curPath))
			return
		}
		process(root.Left, targetSum, curPath)
		process(root.Right, targetSum, curPath)

		// TODO 好奇怪！！有没有下面这段回溯代码都是正确的？？？
		// length := len(curPath)
		// curPath = curPath[:length-1]
		// targetSum += root.Val
	}
	process(root, targetSum, []int{})
	return ans
}

func copySlice(a []int) []int {
	var ans []int
	for _, v := range a {
		ans = append(ans, v)
	}
	return ans
}
